training_scripts/plot_nnunet_training_log.py

```diff
@@ -0,0 +1,157 @@
+        dice_match = dice_pattern.search(line)
+
+        if fold_match:
+            fold_number = int(fold_match.group(1))
+
+        if epoch_match:
+            epoch = int(epoch_match.group(1))
+            epoch_and_dice_data.append({'epoch': epoch, 'pseudo_dice': None})
+
+        elif dice_match:
+            # Extracting the list using regular expression
+            list_match = re.search(r'\[.*\]', dice_match.group())
+
+            if list_match:
+                extracted_list_str = list_match.group(0)
+                # Replace 'nan' with the actual nan value
+                extracted_list_str = extracted_list_str.replace('nan', 'float("nan")')
+                # Evaluating the string representation of the list
+                extracted_list = eval(extracted_list_str)
+                epoch_and_dice_data[-1]['pseudo_dice'] = extracted_list
+
+    # Check if fold_number is instance of int ("if not fold_number:" does not work for '0')
+    if not isinstance(fold_number, int):
+        fold_number = 'all'
+
+    return epoch_and_dice_data, fold_number
+
+
+def main():
+
+    # Parse the command line arguments
+    parser = get_parser()
+    args = parser.parse_args()
+
+    # Get absolute path to the log file
+    log_file_path = os.path.abspath(os.path.expanduser(args.i))
+
+    data, fold_number = extract_epoch_and_dice(log_file_path)
+
+    # Convert data to a Pandas DataFrame
+    df = pd.DataFrame(data)
+
+    # Create columns for each element in pseudo_dice
+    # [:-1] is used to remove the last row, which is empty
+    df_pseudo_dice = pd.DataFrame(df['pseudo_dice'].to_list()[:-1],
+                                  columns=[f'pseudo_dice_{i+1}' for i in range(len(df['pseudo_dice'].iloc[0]))])
+
+    # Concatenate the new DataFrame with the original DataFrame
+    df = pd.concat([df, df_pseudo_dice], axis=1).drop('pseudo_dice', axis=1)
+
+    # Compute mean of pseudo dice across all classes
+    df['pseudo_dice_mean'] = df.iloc[:, 1:].mean(axis=1)
+
+    # Plotting using Plotly Express
+    fig = px.line(df, x='epoch', y=df.columns[1:], title='Pseudo Dice vs. Epoch')
+    # Update the line color for 'pseudo_dice_mean' to black
+    fig.update_traces(line=dict(color='black', width=3), selector=dict(name='pseudo_dice_mean'))
+    # Fix the y-axis range to be between 0 and 1
+    fig.update_yaxes(range=[-0.1, 1.1])
+    # Add y-axis title
+    fig.update_yaxes(title_text='Dice')
+    # Add title with fold number
+    fig.update_layout(title=f'Fold {fold_number} -- Pseudo Dice vs. Epoch (Training)')
+
+    # Show the interactive plot
+    if args.interactive_figure:
+        fig.show()
+
+    # Save plot to a file
+    fname_fig = log_file_path.replace('.txt', '.png')
+    fig.write_image(fname_fig, width=1920, height=1080)
+    print(f'Saved plot to {fname_fig}')
+
+    # Print the latest Dice for each class
+    print(f'Latest Dice for each class: {df.iloc[-2, 1:-1].to_list()}')
```

Comment:
Again, great catch! Thanks!

So, just for the clarification, do I understand the difference correctly? 

- `validation dice` is computed from validation samples (not used for training) after the model training completely finish
- `validation pseudo-dice` is computed on randomly drawn patches from the validation data at the end of each epoch

Reply:
* minor correction: validation dice is computed from validation samples (not used for training) **on full images (not patches)** after the model training completely finish
* minor correction: validation pseudo-dice is computed on randomly drawn patches (**not full images**) from the validation data at the end of each epoch

Why the distinction between patches vs full images is important? because consider a random patch which by chance has 75% of the SC in it. If, in the early training stages, the model predicts the patch correctly then you might think that the model is performing well and might stop training. This shouldn't be done because as the model is still learning and might have correctly predicted the patch by random chance! 

On the other hand, if you do validation on fulll images, then the model _has to_ get the prediction fully correct in order to have a good dice score. Hence, "a big grain of salt" should be taken to understand validation dice and pseudo-validation dice

Reply:
Fixed to `Validation Pseudo Dice` in https://github.com/ivadomed/utilities/pull/41/commits/5241a5a86d48d80e86ef18a0becb1c9c3b615ecd